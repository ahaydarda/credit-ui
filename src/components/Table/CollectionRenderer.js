import React from "react";

export function CollectionRenderer({collection, itemRender, children}) {
    const ui = typeof children === 'function' ? children(collection) : children;
    if(ui ===undefined && !collection ) {
        return null;
    }
    return ui ? ui : (collection && collection.map(itemRender));
}
