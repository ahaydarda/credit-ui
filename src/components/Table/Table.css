.ctable table {
    display: grid;
    border-collapse: collapse;
    min-width: 100%;
    /*grid-template-columns: minmax(150px, 3.33fr)*/
    /*minmax(150px, 3.33fr)*/
    /*minmax(150px, 3.33fr)*/
    border: var(--ctable-border, 0.1em solid lightgray) ;
}

.ctable thead,
.ctable tbody,
.ctable tr {
    display: contents;
}

.ctable  th,
.ctable  td {
    padding: var(--ctable-row-padding, 15px);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    text-align: left;
    border : var(--ctable-row-border,solid 0.5px grey);
}

.ctable th {
    position: sticky;
    top: 0;
    background: var(--ctable-header-background-color, #154C8A) ;
    text-align: left;
    font-weight: normal;
    font-size: var(--ctable-header-font-size, 1.1rem);
    color: var(--ctable-header-color, white) ;
    border : var(--ctable-header-border,solid 0.5px grey);
}

.ctable th:last-child {
    border: 0;
}

.ctable td {
    padding-top: 10px;
    padding-bottom: 10px;
    font-size: var(--ctable-header-font-size, 0.9rem);
    color: var(--ctable-row-color, #808080) ;
}

/*.ctable tr:nth-child(even) td {*/
/*    background: var(--ctable-row-split-color ,#f8f6ff) ;*/
/*}*/

