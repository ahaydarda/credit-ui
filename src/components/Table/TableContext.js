import React from "react";

const TableContext = React.createContext();

export function TableConsumer(props) {
    return (
        <TableContext.Consumer {...props}>
            {context => {
                if (!context) {
                    throw new Error(
                        `Table.Consumer cannot be rendered outside the Table component`,
                    )
                }
                return props.children(context)
            }}
        </TableContext.Consumer>
    )
}

export function TableProvider({value, children}) {
    return (
        <TableContext.Provider value={value} >
            { children }
        </TableContext.Provider>
    );
}
