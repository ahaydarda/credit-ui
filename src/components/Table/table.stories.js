import React from 'react';
import { action } from '@storybook/addon-actions';
import { Table } from './Table';

const {TableBody, TableHeader} = Table;
const { TableTitle} = TableHeader;
const { TableRow} = TableBody;
const { RowItem} = TableRow;

export default {
  title: 'Table',
  component: Table,
};

const prices = [
    {
        "ticker": "ALPHA",
        "price": 3150.67,
        "assetClass": "Macro"
    },
    {
        "ticker": "BETA",
        "price": 3750.67,
        "assetClass": "Macro"
    },
    {
        "ticker": "BETA",
        "price": 3791.37,
        "assetClass": "Equities"
    },
]

 const columns = [
    {
        title: 'Asset Class',
        dataIndex: 'assetClass',
        key: 'assetClass',
        width: 200,
        unit: 4

    },
    {
        title: 'Price',
        dataIndex: 'price',
        key: 'price',
        width: 50,
        unit:3
    },
    {
        title: 'Ticker',
        dataIndex: 'ticker',
        key: 'ticker',
        width: 150,
        unit: 3
    }
];


export const withDirectData = () => <Table columns={columns} data={prices}/>;

export const  renderPropsTable = () =>(
    <Table columns={columns} data={prices}>
        <TableHeader>
            {(columns)=> columns && columns.map(column=>( <TableTitle {...column}/>)) }
        </TableHeader>
        <TableBody>
            {(rows)=> rows && rows.map(row=><TableRow data={row}/>) }
        </TableBody>
    </Table>
)



export const customChildrenTable = () => {
    return (
        <Table>
           <TableHeader>
               <TableTitle
                    title='Asset Class'
                    dataIndex= 'assetClass'
                    key = 'assetClass'
                    width={200}
                    unit ={4}
               />
               <TableTitle
                   title='Price'
                   dataIndex= 'price'
                   key = 'price'
                   width={150}
                   unit ={3}
               />
               <TableTitle
                   title='Ticker'
                   dataIndex= 'ticker'
                   key = 'ticker'
                   width={50}
                   unit ={3}
               />

           </TableHeader>
            <TableBody>
                {
                    prices.map(item =>(
                        <TableRow>
                            <RowItem>{item["assetClass"]}</RowItem>
                            <RowItem>{item["price"]}</RowItem>
                            <RowItem>{item["ticker"]}</RowItem>
                        </TableRow>
                    ))
                }
            </TableBody>
        </Table>
    );
};


