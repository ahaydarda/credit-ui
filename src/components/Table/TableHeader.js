import React from "react";
import {TableConsumer} from "./TableContext";
import {CollectionRenderer} from "./CollectionRenderer";

const Title = ({title,children}) =>{
    return (<th>{ children? children: title}</th>)
}

export const TableHeader = ({children}) =>{
    return (
        <thead>
            <tr>
                <TableConsumer>
                    {
                        ({columns}) => (
                            <CollectionRenderer
                                collection={columns}
                                itemRender={({title}) =>(<Title>{title}</Title>)}
                            >
                                {children}
                            </CollectionRenderer>)
                    }
                </TableConsumer>
            </tr>
        </thead>
    )
}

TableHeader.TableTitle = Title;
