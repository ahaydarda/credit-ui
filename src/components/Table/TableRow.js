import React from "react";
import classnames from "classnames";
import {CollectionRenderer} from "./CollectionRenderer";
import {TableConsumer} from "./TableContext";

function getRowItems({columns,row}) {
    return columns && columns.map(({dataIndex})=> {
        const rowItem =row[dataIndex]
        if( rowItem===undefined){
            throw new Error(`undefined dataIndex, can not find dataIndex:${dataIndex} in row ${row}`);
        }
        return {value: rowItem,dataIndex};
    })
}



const RowItem = ({className,children}) =>{
    return (<td className={className}>{children}</td>)
}


export const TableRow = ({ id, data,className, children}) =>{
    return (<tr key={id} className={className}>
            <TableConsumer>
                {({columns}) => (
                    <CollectionRenderer
                        collection={getRowItems({row:data,columns})}
                        itemRender={({value}) =>(<RowItem>{value}</RowItem>)}
                    >
                        {
                            children ?
                                React.Children.map(children,
                                    (child)=>React.cloneElement(child,{...child.props,className:classnames(className,child.props.className)}))
                                : children
                        }
                    </CollectionRenderer>
                )}
            </TableConsumer>
        </tr>
    );
}


TableRow.TableRowItem = RowItem;
