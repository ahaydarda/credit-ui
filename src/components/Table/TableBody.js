import React from "react";
import {CollectionRenderer} from "./CollectionRenderer";
import {TableConsumer} from "./TableContext";
import {TableRow} from "./TableRow";

export const TableBody = ({children}) =>{
    return (
        <tbody>
            <TableConsumer>
                {({data, uniqueId}) => (
                    <CollectionRenderer
                        collection={data}
                        itemRender={(item) =>(<TableRow key={uniqueId} data={item}/>)}
                    >
                        {children}
                    </CollectionRenderer>
                )}
            </TableConsumer>
        </tbody>
    );
}
