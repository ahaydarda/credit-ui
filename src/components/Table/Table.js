import React, {Component} from "react";
import {TableProvider} from "./TableContext";
import {TableHeader} from "./TableHeader";
import {TableBody} from "./TableBody";
import {TableRow} from "./TableRow";
import classnames from 'classnames'
import {calculateGridTemplateColumns} from './TableUtil'
import './Table.css';

function getColumnsFromPropOrChildren({columns, children}){
    if(!columns || columns.length === 0){
        const tableHeader =  children && React.Children.toArray(children)[0];
        const columnsFromChildren =  tableHeader && tableHeader.props.children ?
            React.Children.toArray(tableHeader.props.children).map(child =>child.props)
            : null;
        if(!columnsFromChildren){
            throw new Error("Empty or null table column data, can not render table");
        }
        return columnsFromChildren;
    } else {
        return columns;
    }
}

export class Table extends Component {
    static  TableHeader = TableHeader;
    static  TableBody = TableBody;
    static  TableRow = TableRow;

    render() {
        const {className,...rest} = this.props;
        const columns = getColumnsFromPropOrChildren({...rest});
        const gridTemplateColumns = calculateGridTemplateColumns(columns);
        const {children} = rest;
        const indexOfId =  columns && columns.findIndex(item=>  item.isUniqueId === true)
        const uniqueId = indexOfId>0? columns[indexOfId].dataIndex : null;
        return (
            <TableProvider value={ {...rest, uniqueId}}>
                <div className={classnames(className, 'ctable')}>
                    <table style={{gridTemplateColumns: gridTemplateColumns}}>
                        {
                            children ? children :
                                (<React.Fragment>
                                    <TableHeader/>
                                    <TableBody/>
                                </React.Fragment>)
                        }
                    </table>
                </div>
            </TableProvider>
        )
    };
}

