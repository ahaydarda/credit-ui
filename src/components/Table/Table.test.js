import React from "react";
import {render, waitForElement, act} from '@testing-library/react'
import {Table} from "./Table";
import renderer from "react-test-renderer";



const columns = [
    {
        title: 'ID',
        dataIndex: 'id',
        key: 'id',
        isUniqueId: true,
        width: 200,
        unit: 4

    },
    {
        title: 'Name',
        dataIndex: 'name',
        key: 'name',
        width: 50,
        unit:6
    },
];

const data = [ {id:1, name:"Ali Haydar Detmer Acar"}];

test('it does not render if column information not provided', async () => {
    expect( ()=>render(<Table />)).toThrowError("Empty or null table column data, can not render table");
});


test('it creates table when column and data provided as property', async () => {
    const table = renderer.create( <Table data={data} columns={columns}/>)
        .toJSON();
    expect(table).toMatchSnapshot();
});



