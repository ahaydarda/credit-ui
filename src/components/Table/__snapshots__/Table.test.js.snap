// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it creates table when column and data provided as property 1`] = `
<div
  className="ctable"
>
  <table
    style={
      Object {
        "gridTemplateColumns": "minmax(200px,4fr) minmax(50px,6fr)",
      }
    }
  >
    <thead>
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          Ali Haydar Detmer Acar
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
